### PR TITLE
add some method for utils

### PR DESCRIPTION
**Pull request by lambda7xx**
add some method for utils

### CODE REVIEW

File: lib/utils/include/utils/graph/README.md
Comment (lockshaw) [inline lib/utils/include/utils/graph/README.md:1]: Why did you delete the README?

File: lib/utils/include/utils/graph/algorithms.h [hunk 1]
```diff
@@ -113,9 +113,11 @@ std::unordered_map<Node, std::unordered_set<Node>>
 std::unordered_map<Node, std::unordered_set<Node>>
     get_predecessors(DiGraphView const &, std::unordered_set<Node> const &);
 
+// return the set of nodes without incoming edges
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/algorithms.h:116]: Change to doxygen syntax

File: lib/utils/include/utils/graph/cow_ptr_t.h
Comment (lockshaw) [inline lib/utils/include/utils/graph/cow_ptr_t.h:1]: Fixed in a separate commit -- pull in changes from the commit linked in #758

File: lib/utils/include/utils/graph/digraph.h [hunk 2]
```diff
@@ -62,7 +62,9 @@ struct DiGraphView {
 
   DiGraphView() = delete;
 
-  operator GraphView() const;
+  operator GraphView() const {
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/digraph.h:65]: Move implementations to `.cc` files to avoid unnecessary compilations

File: lib/utils/include/utils/graph/multidigraph.h [hunk 3]
```diff
@@ -8,37 +8,19 @@
 #include "utils/type_traits.h"
 #include "utils/unique.h"
 #include "utils/visitable.h"
+#include <memory>
 #include <unordered_set>
 
 namespace FlexFlow {
 
-/**
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/multidigraph.h:15]: Why was this deleted?

File: lib/utils/include/utils/graph/node.h [hunk 4]
```diff
@@ -73,9 +73,9 @@ struct GraphView {
     return GraphView(std::make_shared<T>(std::forward<Args>(args)...));
   }
 
-private:
-  GraphView(std::shared_ptr<IGraphView const>);
+  GraphView(std::shared_ptr<IGraphView const> ptr) : ptr(ptr) {}
 
+private:
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/node.h:78]:
```suggestion
```

Comment (lockshaw) [reply lib/utils/include/utils/graph/node.h:78]: It would be great if you could review your PRs before requesting reviews (such as with `git add -p`) just to avoid small fixes like this making it in to the review (if you're already doing this, no problem--we all miss things sometimes :slightly_smiling_face:)

File: lib/utils/include/utils/graph/node.h [hunk 5]
```diff
@@ -85,6 +85,7 @@ static_assert(is_rc_copy_virtual_compliant<IGraphView>::value,
 
 struct IGraph : IGraphView {
   IGraph(IGraph const &) = delete;
+  IGraph() = default;
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/node.h:88]: Why?

File: lib/utils/include/utils/graph/open_graph_interfaces.h [hunk 6]
```diff
@@ -8,9 +8,10 @@ namespace FlexFlow {
 
 struct InputMultiDiEdge : public use_visitable_cmp<InputMultiDiEdge> {
   InputMultiDiEdge() = delete;
-  InputMultiDiEdge(std::pair<std::size_t, std::size_t> const &,
-                   Node const &,
-                   std::size_t const &);
+  InputMultiDiEdge(std::pair<std::size_t, std::size_t> const &uid,
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/open_graph_interfaces.h:11]: Implementations in `.cc` files

File: lib/utils/include/utils/graph/undirected.h [hunk 7]
```diff
@@ -9,7 +9,7 @@
 
 namespace FlexFlow {
 
-struct UndirectedEdge : public use_visitable_cmp<UndirectedEdge> {
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/undirected.h:12]: I generally prefer the explicit `public`--if you have an reason to prefer the other though I'd be happy to be convinced the other way

File: lib/utils/include/utils/graph/node.h (hunk 4 above)
Comment (lockshaw) [reply lib/utils/include/utils/graph/node.h:78]: Also why are we making this public? I don't see a reason why this shouldn't be private.

File: lib/utils/include/utils/graph/undirected.h [hunk 8]
```diff
@@ -85,9 +85,10 @@ struct UndirectedGraphView {
         std::make_shared<T>(std::forward<Args>(args)...));
   }
 
-private:
-  UndirectedGraphView(std::shared_ptr<IUndirectedGraphView const>);
+  UndirectedGraphView(std::shared_ptr<IUndirectedGraphView const> ptr)
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/undirected.h:88]: Why is this being made public?

File: lib/utils/include/utils/graph/undirected.h [hunk 9]
```diff
@@ -113,7 +114,10 @@ struct UndirectedGraph {
 
   UndirectedGraph &operator=(UndirectedGraph);
 
-  operator UndirectedGraphView() const;
+  operator UndirectedGraphView() const {
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/undirected.h:117]: See https://github.com/flexflow/FlexFlow/blob/6ef694251aea7b4c655f95cdbd4fb1e39e4bbcc5/lib/utils/src/graph/multidigraph.cc#L87-L89 for how to implement these

File: lib/utils/include/utils/graph/views.h [hunk 10]
```diff
@@ -100,7 +100,9 @@ struct JoinNodeKey {
 namespace std {
 template <>
 struct hash<::FlexFlow::JoinNodeKey> {
-  std::size_t operator()(::FlexFlow::JoinNodeKey const &) const;
+  std::size_t operator()(::FlexFlow::JoinNodeKey const &key) const {
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/views.h:103]: Instead convert `JoinNodeKey` to being visitable and use `MAKE_VISIT_HASHABLE`

File: lib/utils/include/utils/maybe_owned_ref.h [hunk 11]
```diff
@@ -9,8 +9,8 @@ namespace FlexFlow {
 template <typename T>
 struct maybe_owned_ref {
   maybe_owned_ref() = delete;
-  maybe_owned_ref(T *);
-  maybe_owned_ref(std::shared_ptr<T>);
+  maybe_owned_ref(T *ptr) : _ptr(std::shared_ptr<T>(ptr)){};
```
Comment (lockshaw) [inline lib/utils/include/utils/maybe_owned_ref.h:12]: In this case the pointer should not be owned: change to `_ptr(ptr)`

File: lib/utils/src/graph/algorithms.cc [hunk 12]
```diff
@@ -21,6 +21,10 @@ std::unordered_set<Node> get_nodes(IGraphView const &g) {
   return g.query_nodes({});
 }
 
+std::unordered_set<Node> get_nodes(GraphView const &g) {
+  return g.unsafe()->query_nodes({});
```
Comment (lockshaw) [inline lib/utils/src/graph/algorithms.cc:25]:
```suggestion
  return g.query_nodes({});
```
Also, `get_nodes(IGraphView const &)` should be removed--algorithms should now function on `Graph`s, not `IGraph`s

File: lib/utils/src/graph/algorithms.cc [hunk 13]
```diff
@@ -251,6 +273,28 @@ std::unordered_set<Node> get_sources(DiGraphView const &g) {
   return sources;
 }
 
+std::unordered_set<Node> get_sinks(DiGraphView const &g) {
+  std::unordered_set<Node> dsts;
+  for (Node const &n : get_nodes(g)) {
+    auto outgoing = get_outgoing_edges(g, n);
+    if (outgoing.size() == 0) {
```
Comment (lockshaw) [inline lib/utils/src/graph/algorithms.cc:280]:
```suggestion
    if (get_outdoing_edges(g, n).size() == 0) {
```

File: lib/utils/src/graph/algorithms.cc [hunk 14]
```diff
@@ -251,6 +273,28 @@ std::unordered_set<Node> get_sources(DiGraphView const &g) {
...
+      dsts.insert(n);
+    }
+  }
+  return dsts;
+}
+
+std::unordered_set<Node> get_sinks(MultiDiGraphView const &g) {
+  std::unordered_set<Node> dsts;
```
Comment (lockshaw) [inline lib/utils/src/graph/algorithms.cc:288]: Maybe change to use `filter` from `containers.h`? Would shorten this quite a bit I think

File: lib/utils/src/graph/conversions.cc [hunk 15]
```diff
@@ -193,6 +193,11 @@ std::unique_ptr<IDiGraphView>
   return std::unique_ptr<IDiGraphView>(new ViewMultiDiGraphAsDiGraph{multidi});
 }
 
+// DiGraphView unsafe_view_as_digraph(MultiDiGraphView const &) {
```
Comment (lockshaw) [inline lib/utils/src/graph/conversions.cc:196]: Remove

File: lib/utils/src/graph/digraph.cc [hunk 16]
```diff
@@ -14,6 +15,32 @@ DirectedEdgeQuery::DirectedEdgeQuery(
     tl::optional<std::unordered_set<Node>> const &dsts)
     : srcs(srcs), dsts(dsts) {}
 
+DirectedEdgeQuery query_intersection(DirectedEdgeQuery const &lhs,
+                                     DirectedEdgeQuery const &rhs) {
+  assert(lhs.srcs.has_value() && lhs.dsts.has_value() && rhs.srcs.has_value() &&
+         rhs.dsts.has_value());
+
+  tl::optional<std::unordered_set<Node>> srcs_t1 =
```
Comment (lockshaw) [inline lib/utils/src/graph/digraph.cc:23]:
```suggestion
  optional<std::unordered_set<Node>> srcs_t1 =
```

File: lib/utils/src/graph/digraph.cc [hunk 17]
```diff
@@ -14,6 +15,32 @@ DirectedEdgeQuery::DirectedEdgeQuery(
     tl::optional<std::unordered_set<Node>> const &dsts)
     : srcs(srcs), dsts(dsts) {}
 
+DirectedEdgeQuery query_intersection(DirectedEdgeQuery const &lhs,
+                                     DirectedEdgeQuery const &rhs) {
+  assert(lhs.srcs.has_value() && lhs.dsts.has_value() && rhs.srcs.has_value() &&
```
Comment (lockshaw) [inline lib/utils/src/graph/digraph.cc:20]: This assertion should be removed: see https://github.com/flexflow/FlexFlow/blob/6ef694251aea7b4c655f95cdbd4fb1e39e4bbcc5/lib/utils/src/graph/undirected.cc#L14-L24 or the README

File: lib/utils/src/graph/multidigraph.cc [hunk 18]
```diff
@@ -70,14 +79,38 @@ MultiDiEdgeQuery MultiDiEdgeQuery::all() {
...
+                                    MultiDiEdgeQuery const &rhs) {
+  assert(lhs.srcs.has_value() && lhs.dsts.has_value() && rhs.srcs.has_value() &&
+         rhs.dsts.has_value());
+
+  tl::optional<std::unordered_set<Node>> srcs =
+      intersection(*lhs.srcs, *rhs.srcs);
+  tl::optional<std::unordered_set<Node>> dsts =
+      intersection(*lhs.dsts, *rhs.dsts);
+
+  // TODO, how to set srcIdxs, dstIdxs
```
Comment (lockshaw) [inline lib/utils/src/graph/multidigraph.cc:92]: Let's not merge this change until this TODO has been resolved

File: lib/utils/src/graph/multidigraph.cc [hunk 19]
```diff
@@ -70,14 +79,38 @@ MultiDiEdgeQuery MultiDiEdgeQuery::all() {
...
+
 void swap(MultiDiGraphView &lhs, MultiDiGraphView &rhs) {
   using std::swap;
 
   swap(lhs.ptr, rhs.ptr);
 }
 
+MultiDiGraph::operator MultiDiGraphView() const {
+  std::shared_ptr<IMultiDiGraph const> sharedPtr = ptr.get_shared_ptr();
+  return MultiDiGraphView(sharedPtr);
```
Comment (lockshaw) [inline lib/utils/src/graph/multidigraph.cc:104]:
```suggestion
  return MultiDiGraphView(this->ptr.get_shared_ptr());
```

File: lib/utils/src/graph/serialparallel_internal.h [hunk 20]
```diff
@@ -18,9 +18,16 @@ struct SplitASTNode;
 using SplitAST = mpark::variant<SplitASTNode, Node>;
 
 struct SplitASTNode {
```
Comment (lockshaw) [inline lib/utils/src/graph/serialparallel_internal.h:20]: Move implementation to `.cc` file (also make `SplitASTNode` `visitable`)

File: lib/utils/src/graph/views.cc [hunk 21]
```diff
@@ -117,6 +117,13 @@ Node NodeSource::fresh_node() {
   return result;
 }
 
+JoinNodeKey::JoinNodeKey(Node const &node, LRDirection direction)
+    : node(node), direction(direction) {}
+
+bool JoinNodeKey::operator==(JoinNodeKey const &jnk) const {
```
Comment (lockshaw) [inline lib/utils/src/graph/views.cc:123]: Make `JoinNodeKey` `visitable` and use the automatic `==` generation

File: lib/utils/test/src/test_algorithms.cc [hunk 22]
```diff
@@ -1,165 +1,168 @@
...
-
-  CHECK(ordering.size() == n.size());
-  CHECK_BEFORE(0, 1);
-  CHECK_BEFORE(0, 2);
-  CHECK_BEFORE(1, 5);
-  CHECK_BEFORE(2, 3);
-  CHECK_BEFORE(3, 4);
-  CHECK_BEFORE(4, 5);
-}
+// #include "doctest.h"
```
Comment (lockshaw) [inline lib/utils/test/src/test_algorithms.cc:1]: Why comment out all of the tests?

**Review by lockshaw — CHANGES_REQUESTED** [review]
See individual comments

File: lib/utils/include/utils/graph/algorithms.h (hunk 1 above)
Comment (lambda7xx) [reply lib/utils/include/utils/graph/algorithms.h:116]: Get it
